crates/red_knot_python_semantic/src/types/unpacker.rs

```diff
@@ -1,126 +1,185 @@
+
+                        // TODO: Add diagnostic for length mismatch
+
+                        for (index, ty) in tuple_ty_elements.iter().enumerate() {
+                            if let Some(element_types) = target_types.get_mut(index) {
+                                element_types.push(*ty);
+                            }
+                        }
+                    } else {
+                        let ty = if ty.is_literal_string() {
+                            Type::LiteralString
+                        } else {
+                            ty.iterate(self.db())
+                                .unwrap_with_diagnostic(&self.context, AnyNodeRef::from(target))
+                        };
+                        for target_type in &mut target_types {
+                            target_type.push(ty);
+                        }
                     }
                 }
-                Type::StringLiteral(string_literal_ty) => {
-                    // Deconstruct the string literal to delegate the inference back to the
-                    // tuple type for correct handling of starred expressions. We could go
-                    // further and deconstruct to an array of `StringLiteral` with each
-                    // individual character, instead of just an array of `LiteralString`, but
-                    // there would be a cost and it's not clear that it's worth it.
-                    let value_ty = Type::tuple(
-                        self.db(),
-                        std::iter::repeat(Type::LiteralString)
-                            .take(string_literal_ty.python_len(self.db())),
-                    );
-                    self.unpack(target, value_ty, scope);
-                }
-                _ => {
-                    let value_ty = if value_ty.is_literal_string() {
-                        Type::LiteralString
-                    } else {
-                        value_ty
-                            .iterate(self.db())
-                            .unwrap_with_diagnostic(&self.context, AnyNodeRef::from(target))
+
+                for (index, element) in elts.iter().enumerate() {
+                    let element_ty = match target_types[index].as_slice() {
+                        [] => Type::Unknown,
+                        types => UnionType::from_elements(self.db(), types),
                     };
-                    for element in elts {
-                        self.unpack(element, value_ty, scope);
-                    }
+                    self.unpack_inner(element, element_ty);
                 }
-            },
+            }
             _ => {}
         }
     }
 
+    /// Returns the [`Type`] elements inside the given [`TupleType`] taking into account that there
+    /// can be a starred expression in the `elements`.
+    fn tuple_ty_elements(
+        &mut self,
+        elements: &[ast::Expr],
+        tuple_ty: TupleType<'db>,
+    ) -> Cow<'_, [Type<'db>]> {
```

Comment:
Refactor 2: Move the logic that handles starred expression to a separate method mainly to split the main unpacking function and to avoid being a distraction to the main logic.